### PR TITLE
Astro 1430 push button

### PR DESCRIPTION
**Pull request by NicDominguez**
## Brief Description

Creates `rux-push-button` component in Stencil. Direct port from Lit Element with added event emitter and simplified label prop. 

## JIRA Link

[ASTRO-1430](https://rocketcom.atlassian.net/browse/ASTRO-1430)

## Related Issue

This PR will currently not pass regression due to a bug in the astro-components repo in the push buttons disabled opacity style. [Bug Ticket](https://rocketcom.atlassian.net/browse/ASTRO-1595)

## General Notes

The design of the component was modeled directly off of `rux-switch` which operates the same way since the underling element in the push button is a checkbox with a switch role.

Added Features:
- auto incrementing id similar to other input elements
- available label prop which can be overridden with default slot content (LitElement component had two label props that were not rendered)
- rux-change event emitter to allow for parent to respond

## Motivation and Context

Update to stencil repo. 

## Issues and Limitations

I was not sure whether or not I should treat the push-button as a form element in terms of including a unique id. I chose to do so since I can imagine a use case where an application would use push-buttons as toggles.

The readme has been updated with new auto-generating content but the non-auto-generating content was ported directly from the LitElement component. I believe the readme content will be incorporated into the .mdx story so it is not relevant. 

I also left the basic cypress e2e test as the only e2e test. I am unsure of how we want to approach e2e testing when we are only presenting a collection of stories. I expect that we will have larger discussion on the topic and we can go back and add appropriate tests. 

## Types of changes

-   [ ] Bug fix
-   [X] New feature
-   [ ] Breaking change

## Checklist

-   [X] My code follows the code style of this project.
-   [X] My change requires a change to the documentation.
-   [X] I have added tests to cover my changes.
-   [X] All new and existing tests passed.

### CODE REVIEW

File: cypress/fixtures/example.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (nortonprojects) [inline cypress/fixtures/example.json:1]: Unused - make sure you pulled in the right version of next. This ended up getting deleted out of @markacianfrani's PR

File: docs.json [hunk 2]
```diff
@@ -75324,6 +75324,118 @@
             "dependencies": [],
             "dependencyGraph": {}
         },
+        {
```
Comment (nortonprojects) [inline docs.json:75327]: My inclination is to put in a ticket to delete this file entirely - it look like it's generated during the build?

File: src/components.d.ts [hunk 3]
```diff
@@ -10958,6 +10959,20 @@ export namespace Components {
          */
         "value"?: number;
     }
+    interface RuxPushButton {
```
Comment (nortonprojects) [inline src/components.d.ts:10962]: It looks like this one could be generated too?

File: src/components/rux-push-button/readme.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+- Added `rux_` prefixes and BEM-compatible classes to all `satcom_`-prefixed elements. NOTE: `satcom_` will be removed in a future version
+- Removed prefixed linear gradients
+- Removed prefixed transition
+- Fixed added colon to checked pseudo class (e.g., checked became :checked)
+- Alignment issue fixed on toggle button label
+- Updated to WCAG colors
+- Updated transition speed
+
+<!-- Auto Generated Below -->
```
Comment (nortonprojects) [inline src/components/rux-push-button/readme.md:107]: Probably ought to get rid of this part from the repo - if it's generated we shouldn't need to worry about maintaining it

File: src/components/rux-push-button/rux-push-button.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    handleClick(event: MouseEvent) {
+        event.preventDefault()
+        if (this.disabled) return
+        this.checked = !this.checked
+    }
+
+    @Watch('checked')
+    checkedChanged(checked: boolean) {
+        this.ruxChange.emit({
+            checked: checked,
```
Comment (nortonprojects) [inline src/components/rux-push-button/rux-push-button.tsx:50]: Just a note for syntax - if your property and variable have the same name, you can use shorthand: `this.ruxChange.emit({checked})`

File: src/components/rux-push-button/test/rux-push-button.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import { newSpecPage } from '@stencil/core/testing'
+import { RuxPushButton } from '../rux-push-button'
+
+describe('rux-push-button', () => {
+    it('renders', async () => {
```
Comment (nortonprojects) [inline src/components/rux-push-button/test/rux-push-button.spec.tsx:5]: This one is more of an opinion, but test names *ought* to be written as "should" sentences, i.e. `it('should render the default button')`

File: src/components/rux-push-button/test/rux-push-button.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            html: `<rux-push-button></rux-push-button> <rux-push-button></rux-push-button>`,
+        })
+        const secondPushButton = page.body.querySelectorAll(
+            'rux-push-button'
+        )[1]
+
+        const inputId = secondPushButton.shadowRoot
+            .querySelector('input')
+            .getAttribute('id')
+        expect(inputId).toBe('rux-push-button-2')
```
Comment (nortonprojects) [inline src/components/rux-push-button/test/rux-push-button.spec.tsx:32]: You aren't testing what you're wanting here - the point is to make sure the IDs are unique - compare the two. 

It doesn't matter much here as there aren't likely to be any ID bugs, but it's something to keep in mind for more complicated cases.

File: src/components/rux-push-button/test/rux-push-button.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    })
+
+    it('handles event', async () => {
```
Comment (nortonprojects) [inline src/components/rux-push-button/test/rux-push-button.spec.tsx:35]: Again, what is is this test doing? Should it be looking for a click event, or an `rux-change`?

File: src/components/rux-push-button/test/rux-push-button.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            html: `<rux-push-button></rux-push-button>`,
+        })
+        page.waitForChanges()
+        const handleClick = jest.fn()
+        page.root.addEventListener('click', handleClick)
+        page.root.dispatchEvent(new MouseEvent('click'))
+        expect(handleClick).toHaveBeenCalled()
+    })
+
+    it('does not become checked when disabled', async () => {
```
Comment (nortonprojects) [inline src/components/rux-push-button/test/rux-push-button.spec.tsx:47]: Does it emit an `rux-change` when it's disabled?

**Review by nortonprojects — CHANGES_REQUESTED** [review]
The only "blocking" change is to get rid of `example.json`, though I think the test updates are a good idea

**Review by markacianfrani — APPROVED** [review]
Looks good! Can we add the change event as a storybook action? In your mdx story file:

```
<Meta
    title="Components/Push Button"
    component="rux-push-button"
    argTypes={extractArgTypes('rux-push-button')}
    parameters={{
       actions: {
        handles: [
          'rux-change rux-push-button',
        ],
      },
    }}
/>
```

Also I think if we're going to be using this in a form that it should have a value prop, but I created a new ticket for that.

File: docs.json (hunk 2 above)
Comment (NicDominguez) [reply docs.json:75327]: its been removed, it was already in gitignore but had been committed prior.

File: cypress/fixtures/example.json (hunk 1 above)
Comment (NicDominguez) [reply cypress/fixtures/example.json:1]: It was deleted from next but the merge didn't remove it so I deleted it.

File: src/components.d.ts (hunk 3 above)
Comment (NicDominguez) [reply src/components.d.ts:10962]: Removed, and added to gitignore

Comment (markacianfrani) [reply src/components.d.ts:10962]: Eh, we can't gitignore this. Stencil won't build without it and it doesn't create one if it doesn't exist (a bug). So devs will need to create the file first before being able to do anything which is kind of annoying IMO

Comment (nortonprojects) [reply src/components.d.ts:10962]: 👍 not exactly a high priority to get rid of anyway

File: src/components/rux-push-button/test/rux-push-button.spec.tsx (hunk 8 above)
Comment (NicDominguez) [reply src/components/rux-push-button/test/rux-push-button.spec.tsx:35]: This test will check to confirm the handleClick function was called. It appears stencil docs show testing events is done in e2e testing so I will have to add  it using cypress.

File: src/components/rux-push-button/test/rux-push-button.spec.tsx (hunk 9 above)
Comment (NicDominguez) [reply src/components/rux-push-button/test/rux-push-button.spec.tsx:47]: It does not emit rux-change when disabled as the handle click function returns before changing the checked prop which fires the event. 

Do you mean to ask if rux-change should emit when the button is put in a disabled state even if the checked prop remains the same?

Comment (nortonprojects) [reply src/components/rux-push-button/test/rux-push-button.spec.tsx]: Nope, asking if it tests that. The test shouldn't have any knowledge of the internal mechanics of the system, so it shouldn't know (or care) what order the events are created, as long as they happen when they should, and don't when they shouldn't.

**Review by nortonprojects — APPROVED** [review]

**Pull request merged**